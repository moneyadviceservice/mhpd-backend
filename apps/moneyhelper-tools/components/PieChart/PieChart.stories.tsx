import { StoryFn } from '@storybook/react';
import PieChart from '.';

const StoryProps = {
  title: 'Components/TOOLS/PieChart',
  component: PieChart,
};

const data = [
  {
    name: 'Household bills',
    percentage: 50,
    colour: '#007a91',
  },
  {
    name: 'Living costs',
    percentage: 20,
    colour: '#f56727',
  },
  {
    name: 'Finance & Insurance',
    percentage: 15,
    colour: '#bc0061',
  },
  {
    name: 'Family & Friends',
    percentage: 5,
    colour: '#879a24',
  },
  {
    name: 'Travel',
    percentage: 5,
    colour: '#000d3d',
  },
  {
    name: 'Leisure',
    percentage: 5,
    colour: '#0b4a6d',
  },
];

const Template: StoryFn<typeof PieChart> = (args) => <PieChart {...args} />;

export const Default = Template.bind({});
Default.args = {
  items: data,
};
export const Increase = Template.bind({});
Increase.args = {
  items: data,
  value: 10000,
  title: 'Monthly increase',
  direction: 'increase',
};
export const Decrease = Template.bind({});
Decrease.args = {
  items: data,
  value: 10000,
  title: 'Monthly decrease',
  direction: 'decrease',
};

export default StoryProps;
