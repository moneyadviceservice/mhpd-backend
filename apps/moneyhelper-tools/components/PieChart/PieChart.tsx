import { ReactNode, useMemo } from 'react';
import { twMerge } from 'tailwind-merge';
import NumberFormat from '../../components/NumberFormat';

export type Data = {
  name: ReactNode;
  percentage: number;
  colour: string;
};

type CommonPieChartProps = {
  items: Data[];
  title?: string;
  displayPercentage?: number; // item array index to display
};

type NumberDisplayProps = {
  value: number;
  direction: 'increase' | 'decrease';
};

export type PieChartProps = CommonPieChartProps &
  (NumberDisplayProps | object) &
  object;

const hasNumberDisplay = (
  props: PieChartProps,
): props is CommonPieChartProps & NumberDisplayProps =>
  'value' in props &&
  'direction' in props &&
  typeof props.value === 'number' &&
  typeof props.direction === 'string';

const PieChart = (props: PieChartProps) => {
  const { items, title, displayPercentage } = props;

  const colourPercentCSSValue = useMemo(() => {
    let percentage = 0;
    return items
      .map((type, index) => {
        percentage += type.percentage;

        return `${type.colour} ${
          index === 0 ? 0 + '%' : percentage - type.percentage + '%'
        } ${percentage}%`;
      })
      .join(', ');
  }, [items]);

  // @note after height and width being applied below should
  // also be added to the width increase/decrease container
  const afterWidthHeight = displayPercentage ? 65 : 55;
  const afterClasses = `after:absolute after:bg-white after:rounded-full after:top-2/4 after:left-2/4 after:-translate-y-2/4 after:-translate-x-2/4 ${
    displayPercentage
      ? 'after:h-[65%] after:w-[65%]'
      : 'after:h-[55%] after:w-[55%]'
  }`;

  const isIncreaseOrDecrease = hasNumberDisplay(props);

  return (
    <div className="h-0 pt-[100%] overflow-hidden relative flex-none my-10 w-full">
      <div
        className={twMerge(
          afterClasses,
          'rounded-full w-full h-full absolute left-0 right-0 top-0 bottom-0',
        )}
        style={{
          background: `conic-gradient(${colourPercentCSSValue})`,
        }}
      >
        {isIncreaseOrDecrease ? (
          <div
            className={`absolute left-[50%] top-[50%] translate-x-[-50%] translate-y-[-50%] text-center z-10 w-[${afterWidthHeight}%]`}
          >
            <div className="text-lg font-bold">
              <div className="text-gray-800">{title}</div>
              <NumberFormat direction={props.direction} value={props.value} />
            </div>
          </div>
        ) : null}
        {displayPercentage && (
          <div
            className={`absolute left-[50%] top-[50%] translate-x-[-50%] translate-y-[-50%] text-center z-10 w-[${afterWidthHeight}%]`}
          >
            <span className={`text-[32px]`}>
              {items[displayPercentage].percentage}%
            </span>
          </div>
        )}
      </div>
    </div>
  );
};

export default PieChart;
