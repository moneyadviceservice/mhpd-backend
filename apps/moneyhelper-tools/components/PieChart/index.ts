import PieChart from './PieChart';
export type { Data, PieChartProps } from './PieChart';
export default PieChart;
