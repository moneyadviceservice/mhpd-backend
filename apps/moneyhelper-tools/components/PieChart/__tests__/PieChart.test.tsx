import { render } from '@testing-library/react';
import PieChart from '..';

const data = [
  {
    name: 'Household bills',
    percentage: 50,
    colour: '#007a91',
  },
  {
    name: 'Living costs',
    percentage: 20,
    colour: '#f56727',
  },
  {
    name: 'Finance & Insurance',
    percentage: 15,
    colour: '#bc0061',
  },
  {
    name: 'Family & Friends',
    percentage: 5,
    colour: '#879a24',
  },
  {
    name: 'Travel',
    percentage: 5,
    colour: '#000d3d',
  },
  {
    name: 'Leisure',
    percentage: 5,
    colour: '#0b4a6d',
  },
];

describe('PieChart component', () => {
  it('renders correctly', () => {
    const { container } = render(<PieChart items={data} />);
    expect(container.firstChild).toMatchSnapshot();
  });
});
