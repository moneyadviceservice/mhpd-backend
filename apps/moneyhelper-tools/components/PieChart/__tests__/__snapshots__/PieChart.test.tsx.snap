// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PieChart component renders correctly 1`] = `
<div
  class="h-0 pt-[100%] overflow-hidden relative flex-none my-10 w-full"
>
  <div
    class="after:absolute after:bg-white after:rounded-full after:top-2/4 after:left-2/4 after:-translate-y-2/4 after:-translate-x-2/4 after:h-[55%] after:w-[55%] rounded-full w-full h-full absolute left-0 right-0 top-0 bottom-0"
  />
</div>
`;
